runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java

```diff
@@ -97,12 +97,18 @@
             proxy.resetObjectUnsafe(object);
         }
 
-        // next, if the version is older than what we need
-        // sync.
-        if (object.getVersionUnsafe() < getSnapshotTimestamp()) {
-            proxy.syncObjectUnsafe(proxy.getUnderlyingObject(), getSnapshotTimestamp());
+        // Sync only if we never read the object in this transaction
+        // or if the version is older than what we need
+        OptimisticTransactionalContext oCtxt = (OptimisticTransactionalContext)
+                object.getModifyingContextUnsafe();
+        long snapshotOfThisStream =  streamSnapshots.computeIfAbsent(proxy,
+                k -> k.syncObjectUnsafe(k.getUnderlyingObject(), getSnapshotTimestamp()));
```

Comment:
There is a bit of redundant work in one case. If this is the first time this object is accessed in a transaction, but the underlyingObject version is right, we invoke syncObjectUnsafe unnecessarily.

Why not change the logic to:

if (oCtxt == null || oCtxt != this &&
   proxy.getUnderlyingObject().getVersionUnsafe() < getSnapshotTimestamp() ) {
     long snapshotOfThisStream = ...; 
     if (..getVersionUnsafe() < snapshotOfThisStream) proxy.syncObjectUnsafe(..)
}